File: Daily/sbgo/Leetcode1489-最小生成树的关键边和伪关键边.go
package sbgo
//并查集的创建和方法的设置
type unionFind struct{
	parent,size []int
	Area int	//连通区域大小
}
//init
func newUnionFind(n int) *unionFind {
	parent := make([]int,n)
	size := make([]int,n)
	for i := range parent {
		size[i] = 1
		parent[i] = i
	}
	return &unionFind{parent,size,n}
}

func (uf *unionFind) find(x int) int{
	if uf.parent[x] == x {
		return x
	}
	uf.parent[x] = uf.find(uf.parent[x])
	return uf.parent[x]
}

func (uf *unionFind) union(x,y int)  bool {
	fx := uf.find(x)
	fy := uf.find(y)
	if fx == fy{
		return false
	}
	if uf.size[fx] < uf.size[fy]{
		fx,fy = fy,fx
	}
	uf.parent[fy] = fx
	uf.size[fx] += uf.size[fy]
	uf.Area--
	return true	
}

func findCriticalAndPseudoCriticalEdges(n int, edges [][]int) [][]int {
	m := len(edges)
	//给每条边定位
	for i := 0;i < m;i++{
		edges[i] = append(edges[i],i)
	}
	sort.Slice(edges,func(i,j int) bool {return edges[i][2] < edges[j][2]})
	//复用的部分
	calcu := func(uf *unionFind,ignore int) int {
		v  := 0
		for i,e := range  edges{
			if i != ignore && uf.union(e[0],e[1]){
				v += e[2] 
			}
		}
		if uf.Area > 1{
			return math.MaxInt64
		}
		return v
	}
	//最小构建生成代价
	minVal := calcu(newUnionFind(n),-1)

	var key1,key2 []int
	//遍历每一条边，看他的关键性
	for i,e := range edges{
		if calcu(newUnionFind(n),i) > minVal{
			key1 = append(key1,e[3])
			continue
		}

		uf := newUnionFind(n)
		uf.union(e[0],e[1])
		if e[2] + calcu(uf,i) == minVal{
			key2 = append(key2,e[3])
		}
	}
	return [][]int{key1,key2}
}

File: Leetcode/PY/Leetcode1489-最小生成树的关键边和伪关键边.py
from typing import List
class UnionFind:
    def __init__(self,n:int):
        self.parent = list(range(n))
        self.size = [1]*n
        self.Area = n
    def find(self,x:int)->int:
        if self.parent[x] == x:
            return x
        self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    def union(self,x:int,y:int)->bool:
        fx ,fy = self.find(x),self.find(y)
        if fx == fy:
            return False
        if self.size[fx] < self.size[fy]:
            fx,fy = fy,fx
        self.parent[fy] = fx
        self.size[fx] += self.size[fy]
        self.Area -= 1
        return True
class Solution:
    def findCriticalAndPseudoCriticalEdges(self, n: int, edges: List[List[int]]) -> List[List[int]]:
        m = len(edges)
        for i in range(m):
            edges[i].append(i)
        edges.sort(key = lambda x : x[2])

        uf = UnionFind(n)
        val = 0
        for i in range(m):
            if uf.union(edges[i][0],edges[i][1]):
                val += edges[i][2]
        
        ans = [list(),list()]
        
        for i in range(m):
            uf_cur = UnionFind(n)
            vla_cur = 0
            for j in range(m):
                if i != j and uf_cur.union(edges[j][0],edges[j][1]):
                    vla_cur += edges[j][2]

            if  uf_cur.Area != 1 or (uf_cur.Area == 1 and vla_cur > val):
                ans[0].append(edges[i][3])
                continue
            
            uf_cur = UnionFind(n)
            uf_cur.union(edges[i][0],edges[i][1])
            vla_cur = edges[i][2]
            for j in range(m):
                if i != j and  uf_cur.union(edges[j][0],edges[j][1]):
                    vla_cur += edges[j][2]
            
            if val == vla_cur:
                ans[1].append(edges[i][3])

        return ans 